tests/two_d_array_ut/two_d_array_ut.c

```diff
@@ -0,0 +1,642 @@
+
+/* Tests_SRS_TWO_D_ARRAY_07_007: [ If two_d_array is NULL, TWO_D_ARRAY_FREE(T) shall do nothing. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_do_nothing)
+{
+    //arrange
+    TWO_D_ARRAY(uint32_t) tdarr = NULL;
+
+    //act
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+
+    //assert
+    ASSERT_IS_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_009: [ TWO_D_ARRAY_FREE(T) shall free the memory associated with TWO_D_ARRAY(T). ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_succeeds_without_row_allocated)
+{
+    //arrange
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(10000000, 5000);
+    ASSERT_IS_NOT_NULL(tdarr);
+    umock_c_reset_all_calls();
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+
+    //act
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+
+    //assert
+    ASSERT_IS_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_008: [ TWO_D_ARRAY_FREE(T) shall free all rows that are non-NULL. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_succeeds_with_rows_allocated)
+{
+    //arrange
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(10000000, 5000);
+    int add_row_result_0 = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+    int add_row_result_70 = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 70);
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result_0, 0);
+    ASSERT_ARE_EQUAL(int, add_row_result_70, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[70]);
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(free(IGNORED_ARG));
+
+    //act
+    TWO_D_ARRAY_ASSIGN(uint32_t)(&tdarr, NULL);
+
+    //assert
+    ASSERT_IS_NULL(tdarr);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* TWO_D_ARRAY_FREE_ROW(T) */
+
+/* Tests_SRS_TWO_D_ARRAY_07_010: [ If two_d_array is NULL, TWO_D_ARRAY_FREE_ROW(T) shall fail return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_ROW_fails_when_handle_is_null)
+{
+    //arrange
+    int result;
+
+    //act
+    result = TWO_D_ARRAY_FREE_ROW(uint32_t)(NULL, 0);
+
+    //assert
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+
+    //clean
+}
+
+/* Tests_SRS_TWO_D_ARRAY_07_011: [ If row_index is equal or greater than row_size, TWO_D_ARRAY_FREE_ROW(T) shall fail and return a non-zero value. ]*/
+TEST_FUNCTION(TWO_D_ARRAY_FREE_ROW_fails_when_row_index_exceeds)
+{
+    //arrange
+    int result;
+    TWO_D_ARRAY(uint32_t) tdarr = TWO_D_ARRAY_CREATE(uint32_t)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(uint32_t)(tdarr, 0);
+    ASSERT_IS_NOT_NULL(tdarr);
```

Comment:
This assert has to be before we try to use the tdarr
Here and everywhere else where applicable.

Reply:
modified the order here and everywhere else